mlflow/langchain/_langchain_autolog.py

```diff
@@ -0,0 +1,145 @@
+import warnings
+from copy import deepcopy
+
+import langchain
+from packaging.version import Version
+from pydantic.v1.config import Extra
+
+import mlflow
+from mlflow.ml_package_versions import _ML_PACKAGE_VERSIONS
+from mlflow.utils.autologging_utils import (
+    ExceptionSafeAbstractClass,
+    disable_autologging,
+    get_autologging_config,
+)
+from mlflow.utils.autologging_utils.safety import _resolve_extra_tags
+
+MIN_REQ_VERSION = Version(_ML_PACKAGE_VERSIONS["langchain"]["autologging"]["minimum"])
+MAX_REQ_VERSION = Version(_ML_PACKAGE_VERSIONS["langchain"]["autologging"]["maximum"])
+
+_lc_version = Version(langchain.__version__)
+_logger = logging.getLogger(__name__)
+
+
+def _get_input_data_from_invoke(model, args, kwargs):
+    from langchain.schema.runnable import Runnable
+
+    invoke_func = model.invoke
+    input_example_exc = None
+    try:
+        if isinstance(model, Runnable):
+            # A guard to make sure `input` is the first argument of `invoke` function
+            assert next(iter(inspect.signature(invoke_func).parameters.keys())) == "input"
+            return kwargs["input"] if "input" in kwargs else args[0]
+    except Exception as e:
+        input_example_exc = e
+    _logger.warning(
+        f"Failed to gather input example of model {model.__class__.__name__}"
+        + f" due to {input_example_exc}."
+        if input_example_exc
+        else ""
+    )
+
+
+def patched_invoke(original, self, *args, **kwargs):
+    """
+    A patched implementation of langchain runnables `invoke` which enables logging the
+    following parameters, metrics and artifacts:
+
+    - model
+    - model parameters
+    - invoke data
+    """
+    from langchain.callbacks import MlflowCallbackHandler
+    from langchain.schema.runnable.config import RunnableConfig
+
+    class _MLflowLangchainCallback(MlflowCallbackHandler, metaclass=ExceptionSafeAbstractClass):
+        """
+        Callback for auto-logging metrics and parameters.
+        We need to inherit ExceptionSafeAbstractClass to avoid invalid new
+        input arguments added to original function call.
+        """
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+        # def flush_tracker(self, langchain_asset: Any = None, finish: bool = False) -> None:
+        #     return super().flush_tracker(langchain_asset, finish)
+
+    if not MIN_REQ_VERSION <= _lc_version <= MAX_REQ_VERSION:
+        warnings.warn(
+            "Autologging is known to be compatible with langchain versions between "
+            f"{MIN_REQ_VERSION} and {MAX_REQ_VERSION} and may not succeed with packages "
+            "outside this range."
+        )
+
+    config = kwargs.get("config", None)
+    run_id = self.run_id if hasattr(self, "run_id") else None
+    if active_run := mlflow.active_run():
+        run_id = active_run.info.run_id
+        # mlflow callback internally starts the run, so we end it here
+        mlflow.end_run()
```

Comment:
Can we validate the functionality for child runs here? Does the callback work with a child run? Will a parent run still be active and what does that do for the langchain callback if this implementation is active?

Reply:
We could support nested run by checking if there's an active run, then we assume we should create child runs by default. But since we're always just creating a single run, is there any use cases for nested runs? Unless the customer explicitly log another child run, otherwise the parent run will always contain a single child run.